### ISSUE TITLE
Different Behaviour in node v9.2.0 & v10.0.0

### ISSUE THREAD

**Issue by DeusProx** [nodejs/abi-stable-node#300]
I'm using NAPI to write an object wrapper for C++ code. For learning & prototyping I used `assert(status != napi_ok)` to check if everything works correct but atm I'm switching to `if(status != napi_ok) napi_throw_error(...);` but forgot to return directly after `napi_throw_error`.
Due to this I realised that next function `napi_unwrap` returns differently in node v9.2.0 and node v10.0.0 resulting in failing the following assert in node v10.0.0 but not in v9.2.0 when the function is called with a parameter (E.g. `tree.print(1)` in JS). Without parameters everything works as expected.

It's happening in following function:
```Cpp
napi_value
Tree::NapiPrint(napi_env env, napi_callback_info info) {
    napi_status status;
    napi_value self;
    size_t argc = 0;

    Tree* tree;
    std::string treeString;
    napi_value outputTreeString;

    // get callback info
    status = napi_get_cb_info(env, info, &argc, NULL, &self, NULL);
    if (status != napi_ok || argc != 0) {
        napi_throw_error(env, "100",  "Wrong number of arguments. 0 arguments expected.");
        // forgot to return here
    }

    // unwrap
    status = napi_unwrap(env, self, reinterpret_cast<void**>(&tree));
    assert(status == napi_ok); // This fails in node v10.0.0

    // generate napi string of tree structure
    treeString = tree->print();
    status = napi_create_string_utf8(env, treeString.c_str(), treeString.length(), &outputTreeString);
    assert(status == napi_ok);

    return outputTreeString;
}
```

The solution for me seems pretty clear: Just return as the other parts of the code shouldn't be executed. But I don't know if this change in behaviour between different node versions is of interest. 

If not please just close this issue!

**Comment by gabrielschulhof** [comment]
@DeusProx from 9.2.0 to 10.0.0 we changed the implementation of `napi_wrap` in order to make it less invasive on the JavaScript object. Whereas before we were manipulating an object's prototype chain to store the native data, we switched to using private properties for the same purpose.

Setting private properties has the same semantics as setting regular properties, meaning that JavaScript may be executed as a result. Thus, exceptions may be thrown as a result. This means that ever since we switched to using private properties for storing native pointers we need to first check whether an exception is already pending before we attempt to retrieve the native pointer because retrieving it may cause another exception to be thrown, which would cause the first exception to be silently ignored.

Thus, the exception you throw in your addon is preventing the unwrap from succeeding in 10.0.0, but not in 9.2.0.
